File: go/extractor.go
package headgate

// Typed handler extractors. Extraction is explicit manual DI: dependencies enter at
// runner construction through Config.Extensions, never through a global container.
// Every extractor runs after payload decoding and before the user's handler function.

import (
	"context"
	"fmt"
	"reflect"
)

// ExtractionError reports which pre-handler extractor failed. It travels through the
// normal attempt error path, but the registered user function is never entered.
type ExtractionError struct {
	Extractor string
	Message   string
}

func (e *ExtractionError) Error() string {
	return fmt.Sprintf("extract %s: %s", e.Extractor, e.Message)
}

// HandlerExtractor constructs one typed handler parameter from the dispatch context.
// Applications can implement it directly; ExtractorFunc is the adapter for a function.
type HandlerExtractor[T any] interface {
	Extract(context.Context) (T, error)
}

type ExtractorFunc[T any] func(context.Context) (T, error)

func (f ExtractorFunc[T]) Extract(ctx context.Context) (T, error) { return f(ctx) }

// Metadata is the durable, non-payload envelope metadata visible at dispatch.
type Metadata struct {
	Queue, PartitionKey, RateClass string
	Weight                         uint32
	Priority                       int32
	SchemaVersion                  uint32
	Headers                        map[string]string
}

// Attempt keeps returned errors and crash-attributed losses distinct.
type Attempt struct {
	ReturnedErrors uint32
	Crashes        uint32
	MaxAttempts    uint32
}

type TaskID string

// WorkerContext contains stable facts about the runner, not its dependency container.
type WorkerContext struct {
	WorkerID string
	Queues   []string
	Capacity int
}

type extractionScope struct {
	metadata Metadata
	attempt  Attempt
	taskID   TaskID
	worker   WorkerContext
}

type extractionContextKey struct{}

func withExtractionScope(ctx context.Context, claim Claim, worker WorkerContext) context.Context {
	headers := make(map[string]string, len(claim.Envelope.Headers))
	for key, value := range claim.Envelope.Headers {
		headers[key] = value
	}
	worker.Queues = append([]string(nil), worker.Queues...)
	return context.WithValue(ctx, extractionContextKey{}, extractionScope{
		metadata: Metadata{
			Queue: claim.Envelope.Queue, PartitionKey: claim.Envelope.PartitionKey,
			RateClass: claim.Envelope.RateClass, Weight: EffectiveWeight(claim.Envelope.Weight),
			Priority: claim.Envelope.Priority, SchemaVersion: claim.Envelope.SchemaVersion,
			Headers: headers,
		},
		attempt: Attempt{
			ReturnedErrors: claim.Envelope.Attempt, Crashes: claim.Envelope.CrashAttempt,
			MaxAttempts: claim.Envelope.MaxAttempts,
		},
		taskID: TaskID(claim.Envelope.ID),
		worker: worker,
	})
}

func extractionScopeFrom(ctx context.Context) (extractionScope, bool) {
	scope, ok := ctx.Value(extractionContextKey{}).(extractionScope)
	return scope, ok
}

func unavailableExtractor[T any](name string) (T, error) {
	var zero T
	return zero, &ExtractionError{Extractor: name, Message: "dispatch context is unavailable"}
}

// ExtractData resolves T from job data first, then worker data. Asking for the wrong
// concrete type is a missing-data error, never an untyped cast inside the handler.
func ExtractData[T any]() HandlerExtractor[T] {
	return ExtractorFunc[T](func(ctx context.Context) (T, error) {
		if value, ok := Data[T](ctx); ok {
			return value, nil
		}
		var zero T
		return zero, &ExtractionError{
			Extractor: "Data",
			Message:   fmt.Sprintf("missing typed data `%s`", extensionType[T]()),
		}
	})
}

func ExtractMetadata() HandlerExtractor[Metadata] {
	return ExtractorFunc[Metadata](func(ctx context.Context) (Metadata, error) {
		scope, ok := extractionScopeFrom(ctx)
		if !ok {
			return unavailableExtractor[Metadata]("Metadata")
		}
		metadata := scope.metadata
		metadata.Headers = make(map[string]string, len(scope.metadata.Headers))
		for key, value := range scope.metadata.Headers {
			metadata.Headers[key] = value
		}
		return metadata, nil
	})
}

// ExtractMeta validates the durable metadata into an application type before the
// handler runs. A missing or malformed header should be returned by decode as an error.
func ExtractMeta[T any](decode func(Metadata) (T, error)) HandlerExtractor[T] {
	return ExtractorFunc[T](func(ctx context.Context) (T, error) {
		metadata, err := ExtractMetadata().Extract(ctx)
		if err != nil {
			var zero T
			return zero, err
		}
		value, err := decode(metadata)
		if err != nil {
			var zero T
			return zero, &ExtractionError{
				Extractor: "Meta",
				Message:   fmt.Sprintf("%s: %v", reflect.TypeOf((*T)(nil)).Elem(), err),
			}
		}
		return value, nil
	})
}

func ExtractAttempt() HandlerExtractor[Attempt] {
	return ExtractorFunc[Attempt](func(ctx context.Context) (Attempt, error) {
		scope, ok := extractionScopeFrom(ctx)
		if !ok {
			return unavailableExtractor[Attempt]("Attempt")
		}
		return scope.attempt, nil
	})
}

func ExtractTaskID() HandlerExtractor[TaskID] {
	return ExtractorFunc[TaskID](func(ctx context.Context) (TaskID, error) {
		scope, ok := extractionScopeFrom(ctx)
		if !ok {
			return unavailableExtractor[TaskID]("TaskId")
		}
		return scope.taskID, nil
	})
}

func ExtractWorkerContext() HandlerExtractor[WorkerContext] {
	return ExtractorFunc[WorkerContext](func(ctx context.Context) (WorkerContext, error) {
		scope, ok := extractionScopeFrom(ctx)
		if !ok {
			return unavailableExtractor[WorkerContext]("WorkerContext")
		}
		worker := scope.worker
		worker.Queues = append([]string(nil), worker.Queues...)
		return worker, nil
	})
}

func ExtractClient() HandlerExtractor[*JobClient] {
	return ExtractorFunc[*JobClient](func(ctx context.Context) (*JobClient, error) {
		client, ok := ClientFromContext(ctx)
		if !ok {
			return nil, ErrClientFromContextUnavailable
		}
		return client, nil
	})
}

// RegisterExtractedN keeps extraction compile-time typed without reflection or a
// service locator. All N extractors finish before work is called.
func RegisterExtracted1[T Args, A any](r *Registry, a HandlerExtractor[A], work func(context.Context, *Job[T], A) error) error {
	return RegisterFunc[T](r, func(ctx context.Context, job *Job[T]) error {
		av, err := a.Extract(ctx)
		if err != nil {
			return err
		}
		return work(ctx, job, av)
	})
}

func RegisterExtracted2[T Args, A, B any](r *Registry, a HandlerExtractor[A], b HandlerExtractor[B], work func(context.Context, *Job[T], A, B) error) error {
	return RegisterFunc[T](r, func(ctx context.Context, job *Job[T]) error {
		av, err := a.Extract(ctx)
		if err != nil {
			return err
		}
		bv, err := b.Extract(ctx)
		if err != nil {
			return err
		}
		return work(ctx, job, av, bv)
	})
}

func RegisterExtracted3[T Args, A, B, C any](r *Registry, a HandlerExtractor[A], b HandlerExtractor[B], c HandlerExtractor[C], work func(context.Context, *Job[T], A, B, C) error) error {
	return RegisterFunc[T](r, func(ctx context.Context, job *Job[T]) error {
		av, err := a.Extract(ctx)
		if err != nil {
			return err
		}
		bv, err := b.Extract(ctx)
		if err != nil {
			return err
		}
		cv, err := c.Extract(ctx)
		if err != nil {
			return err
		}
		return work(ctx, job, av, bv, cv)
	})
}

func RegisterExtracted4[T Args, A, B, C, D any](r *Registry, a HandlerExtractor[A], b HandlerExtractor[B], c HandlerExtractor[C], d HandlerExtractor[D], work func(context.Context, *Job[T], A, B, C, D) error) error {
	return RegisterFunc[T](r, func(ctx context.Context, job *Job[T]) error {
		av, err := a.Extract(ctx)
		if err != nil {
			return err
		}
		bv, err := b.Extract(ctx)
		if err != nil {
			return err
		}
		cv, err := c.Extract(ctx)
		if err != nil {
			return err
		}
		dv, err := d.Extract(ctx)
		if err != nil {
			return err
		}
		return work(ctx, job, av, bv, cv, dv)
	})
}

func RegisterExtracted5[T Args, A, B, C, D, E any](r *Registry, a HandlerExtractor[A], b HandlerExtractor[B], c HandlerExtractor[C], d HandlerExtractor[D], e HandlerExtractor[E], work func(context.Context, *Job[T], A, B, C, D, E) error) error {
	return RegisterFunc[T](r, func(ctx context.Context, job *Job[T]) error {
		av, err := a.Extract(ctx)
		if err != nil {
			return err
		}
		bv, err := b.Extract(ctx)
		if err != nil {
			return err
		}
		cv, err := c.Extract(ctx)
		if err != nil {
			return err
		}
		dv, err := d.Extract(ctx)
		if err != nil {
			return err
		}
		ev, err := e.Extract(ctx)
		if err != nil {
			return err
		}
		return work(ctx, job, av, bv, cv, dv, ev)
	})
}

File: crates/headgate/src/extractor.rs
//! Typed handler extractors.
//!
//! Extraction happens after payload decoding but before the user's handler is called.
//! An extraction failure is therefore an ordinary attempt error with no handler side
//! effects. This is intentionally manual, compile-time DI rather than a global service
//! locator: worker dependencies enter through [`crate::WorkerConfig::extensions`].

use std::collections::BTreeMap;
use std::ops::Deref;
use std::sync::Arc;

use headgate_core::Envelope;

use crate::{JobClient, JobCtx};

/// A typed extraction failure. The message names the failed extractor and is returned
/// through the normal attempt error path; the handler itself is never entered.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ExtractionError {
    extractor: &'static str,
    message: String,
}

impl ExtractionError {
    pub fn new(extractor: &'static str, message: impl Into<String>) -> Self {
        Self {
            extractor,
            message: message.into(),
        }
    }

    pub fn extractor(&self) -> &'static str {
        self.extractor
    }

    pub fn message(&self) -> &str {
        &self.message
    }
}

impl std::fmt::Display for ExtractionError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "extract {}: {}", self.extractor, self.message)
    }
}

impl std::error::Error for ExtractionError {}

/// One value that can be built from the dispatch context before a handler runs.
pub trait FromJobRequest: Sized + Send + 'static {
    fn from_job(ctx: &JobCtx, envelope: &Envelope) -> Result<Self, ExtractionError>;
}

/// Typed process-local data. Job-local `T` shadows the worker's `T`, matching
/// [`JobCtx::data`].
#[derive(Clone, Debug)]
pub struct Data<T>(pub Arc<T>);

impl<T> Deref for Data<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> FromJobRequest for Data<T>
where
    T: Send + Sync + 'static,
{
    fn from_job(ctx: &JobCtx, _: &Envelope) -> Result<Self, ExtractionError> {
        ctx.data::<T>().map(Self).ok_or_else(|| {
            ExtractionError::new(
                "Data",
                format!("missing typed data `{}`", std::any::type_name::<T>()),
            )
        })
    }
}

/// The durable, non-payload metadata visible at dispatch. Header values stay strings;
/// use [`Meta<T>`] when an application wants validated typed metadata.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Metadata {
    pub queue: String,
    pub partition_key: String,
    pub rate_class: String,
    pub weight: u32,
    pub priority: i32,
    pub schema_version: u32,
    pub headers: BTreeMap<String, String>,
}

impl Metadata {
    fn from_envelope(envelope: &Envelope) -> Self {
        Self {
            queue: envelope.queue.clone(),
            partition_key: envelope.partition_key.clone(),
            rate_class: envelope.rate_class.clone(),
            weight: headgate_core::effective_weight(envelope.weight),
            priority: envelope.priority,
            schema_version: envelope.schema_version,
            headers: envelope.headers.clone(),
        }
    }
}

impl FromJobRequest for Metadata {
    fn from_job(_: &JobCtx, envelope: &Envelope) -> Result<Self, ExtractionError> {
        Ok(Self::from_envelope(envelope))
    }
}

/// Application-defined conversion from the durable metadata snapshot into a validated
/// type. This keeps parsing out of the handler and makes a malformed/missing value fail
/// before handler side effects.
pub trait FromMetadata: Sized + Send + 'static {
    fn from_metadata(metadata: &Metadata) -> Result<Self, String>;
}

#[derive(Clone, Debug)]
pub struct Meta<T>(pub T);

impl<T> Deref for Meta<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> FromJobRequest for Meta<T>
where
    T: FromMetadata,
{
    fn from_job(_: &JobCtx, envelope: &Envelope) -> Result<Self, ExtractionError> {
        let metadata = Metadata::from_envelope(envelope);
        T::from_metadata(&metadata).map(Self).map_err(|message| {
            ExtractionError::new("Meta", format!("{}: {message}", std::any::type_name::<T>()))
        })
    }
}

/// Returned errors, crash-attributed losses, and retry budget at this delivery.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Attempt {
    pub returned_errors: u32,
    pub crashes: u32,
    pub max_attempts: u32,
}

impl FromJobRequest for Attempt {
    fn from_job(_: &JobCtx, envelope: &Envelope) -> Result<Self, ExtractionError> {
        Ok(Self {
            returned_errors: envelope.attempt,
            crashes: envelope.crash_attempt,
            max_attempts: envelope.max_attempts,
        })
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct TaskId(pub String);

impl Deref for TaskId {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl FromJobRequest for TaskId {
    fn from_job(_: &JobCtx, envelope: &Envelope) -> Result<Self, ExtractionError> {
        Ok(Self(envelope.id.clone()))
    }
}

/// Stable facts about the runner executing this attempt. It contains no dependency
/// container: dependencies are extracted explicitly through [`Data<T>`].
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct WorkerContext {
    worker_id: String,
    queues: Arc<[String]>,
    capacity: u32,
}

impl WorkerContext {
    pub(crate) fn new(worker_id: String, queues: Vec<String>, capacity: u32) -> Self {
        Self {
            worker_id,
            queues: queues.into(),
            capacity,
        }
    }

    pub fn worker_id(&self) -> &str {
        &self.worker_id
    }

    pub fn queues(&self) -> &[String] {
        &self.queues
    }

    pub fn capacity(&self) -> u32 {
        self.capacity
    }
}

impl FromJobRequest for WorkerContext {
    fn from_job(ctx: &JobCtx, _: &Envelope) -> Result<Self, ExtractionError> {
        Ok(ctx.worker_context().clone())
    }
}

impl FromJobRequest for JobClient {
    fn from_job(ctx: &JobCtx, _: &Envelope) -> Result<Self, ExtractionError> {
        Ok(ctx.client().clone())
    }
}

/// A tuple of extractors. Implemented up to arity eight; the handler receives the tuple
/// and can destructure it in its argument pattern.
pub trait FromJobRequestTuple: Sized + Send + 'static {
    fn from_job(ctx: &JobCtx, envelope: &Envelope) -> Result<Self, ExtractionError>;
}

impl FromJobRequestTuple for () {
    fn from_job(_: &JobCtx, _: &Envelope) -> Result<Self, ExtractionError> {
        Ok(())
    }
}

macro_rules! tuple_extractors {
    ($(($name:ident, $index:tt)),+ $(,)?) => {
        impl<$($name),+> FromJobRequestTuple for ($($name,)+)
        where
            $($name: FromJobRequest,)+
        {
            fn from_job(ctx: &JobCtx, envelope: &Envelope) -> Result<Self, ExtractionError> {
                Ok(($($name::from_job(ctx, envelope)?,)+))
            }
        }
    };
}

tuple_extractors!((E1, 0));
tuple_extractors!((E1, 0), (E2, 1));
tuple_extractors!((E1, 0), (E2, 1), (E3, 2));
tuple_extractors!((E1, 0), (E2, 1), (E3, 2), (E4, 3));
tuple_extractors!((E1, 0), (E2, 1), (E3, 2), (E4, 3), (E5, 4));
tuple_extractors!((E1, 0), (E2, 1), (E3, 2), (E4, 3), (E5, 4), (E6, 5));
tuple_extractors!(
    (E1, 0),
    (E2, 1),
    (E3, 2),
    (E4, 3),
    (E5, 4),
    (E6, 5),
    (E7, 6)
);
tuple_extractors!(
    (E1, 0),
    (E2, 1),
    (E3, 2),
    (E4, 3),
    (E5, 4),
    (E6, 5),
    (E7, 6),
    (E8, 7)
);
